crates/chain/src/indexed_tx_graph.rs

```diff
@@ -0,0 +1,406 @@
+use core::{convert::Infallible, ops::AddAssign};
+
+use bitcoin::{OutPoint, Script, Transaction, TxOut};
+
+use crate::{
+    keychain::Balance,
+    sparse_chain::ChainPosition,
+    tx_graph::{Additions, TxGraph, TxNode},
+    BlockAnchor, ChainOracle, FullTxOut, ObservedIn, TxIndex,
+};
+
+/// An outwards-facing view of a transaction that is part of the *best chain*'s history.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxInChain<'a, T, A> {
+    /// Where the transaction is observed (in a block or in mempool).
+    pub observed_in: ObservedIn<&'a A>,
+    /// The transaction with anchors and last seen timestamp.
+    pub tx: TxNode<'a, T, A>,
+}
+
+/// A structure that represents changes to an [`IndexedTxGraph`].
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "A: Ord + serde::Deserialize<'de>, IA: serde::Deserialize<'de>",
+            serialize = "A: Ord + serde::Serialize, IA: serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct IndexedAdditions<A, IA> {
+    /// [`TxGraph`] additions.
+    pub graph_additions: Additions<A>,
+    /// [`TxIndex`] additions.
+    pub index_additions: IA,
+    /// Last block height witnessed (if any).
+    pub last_height: Option<u32>,
+}
+
+impl<A, IA: Default> Default for IndexedAdditions<A, IA> {
+    fn default() -> Self {
+        Self {
+            graph_additions: Default::default(),
+            index_additions: Default::default(),
+            last_height: None,
+        }
+    }
+}
+
+impl<A: BlockAnchor, IA: AddAssign> AddAssign for IndexedAdditions<A, IA> {
+    fn add_assign(&mut self, rhs: Self) {
+        let Self {
+            graph_additions,
+            index_additions: index_delta,
+            last_height,
+        } = rhs;
+        self.graph_additions.append(graph_additions);
+        self.index_additions += index_delta;
+        if self.last_height < last_height {
+            let last_height =
+                last_height.expect("must exist as it is larger than self.last_height");
+            self.last_height.replace(last_height);
+        }
+    }
+}
+
+pub struct IndexedTxGraph<A, I> {
+    /// Transaction index.
+    pub index: I,
+    graph: TxGraph<A>,
+    last_height: u32,
```

Comment:
I think it adds complexity, we can just assume the default last-seen height is 0 (genesis block).

Reply:
why does this even exist at all?

Reply:
If you're sure this doesn't add any off-by-one bugs, as you're basically always saying that you already saw the genesis block, then alright.

Reply:
> why does this even exist at all?

Because chain state and tx graph are no longer updated together, so we need to keep track of height that we are synced up to in the tx graph side.

Reply:
> If you're sure this doesn't add any off-by-one bugs, as you're basically always saying that you already saw the genesis block, then alright.

I'm pretty sure. This is only used to tell the "chain source" where to start syncing from. There are no important transactions in the genesis block anyway. Also, we typically want to sync from `tip - assume_final_depth` as well.

Reply:
> > why does this even exist at all?
> 
> Because chain state and tx graph are no longer updated together, so we need to keep track of height that we are synced up to in the tx graph side.

Why do we need to do this?

Reply:
@LLFourn If want to start syncing from a block-by-block chain source, we can then say, start from this specific height.

Reply:
So doesn't it make sense for the block-by-block chain source to keep track of this? Why does a txgraph care about this?

Reply:
The benefit of keeping the `last_seen_height` on the `bdk_chain` side is that we can guarantee the `last_seen_height` and the chain data can be "updated atomically". This has two benefits that I can think of.

1. Syncing from multiple sources (This maybe a niche use case, but it will be nice to able to change chain source and not have to sync from genesis).

2. Persistence. Ideally, we only want to persist `last_seen_height` with the chain data introduced when syncing up to that height. However, I guess we can make the chain source do persistence as well, but I think it's more elegant if the chain source doesn't require persisting state.

Reply:
> @LLFourn If want to start syncing from a block-by-block chain source, we can then say, start from this specific height.

what about if there has been a re-org since you recorded that. It is not sufficient to start at that height anymore right? This means it should be a list of block hashes and heights (go backwards until you find one that it currently in chain and start from there).

I get that persisting stuff about your blockchain source history is useful and we should make it easy to persist but I don't think it belongs on `TxGraph`. In `Wallet` for example this will happen naturally since it uses a `LocalChainOracle` internally. For other systems we can make it easy to attach some application data to persistence store. This is a separate issue though.

Reply:
This has been removed now.